x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/NGramAutomaton.java

```diff
@@ -0,0 +1,374 @@
+    }
+
+    /**
+     * Convert this automaton into an expression of ngrams that must be found
+     * for the entire automaton to match. The automaton isn't simplified so you
+     * probably should call {@link Expression#simplify()} on it.
+     */
+    public Expression<String> expression() {
+        return Or.fromExpressionSources(acceptStates);
+    }
+
+    /**
+     * Recursively walk transitions building the prefixes for the initial state.
+     *
+     * @param codePoints work array holding codePoints
+     * @param offset offset into work array/depth in tree
+     * @param currentState current source state
+     * @return true to continue, false if we hit a dead end
+     */
+    private boolean buildInitial(int[] codePoints, int offset, int currentState) {
+        if (source.isAccept(currentState)) {
+            // Hit an accept state before finishing a trigram - meaning you
+            // could match this without using any of the trigrams we might find
+            // later. In that case we just give up.
+            initialStates.clear();
+            states.clear();
+            return false;
+        }
+        if (offset == gramSize - 1) {
+            // We've walked deeply enough to find an initial state.
+            NGramState state = new NGramState(currentState, new String(codePoints, 0, gramSize - 1), true);
+            // Only add one copy of each state - if we've already seen this
+            // state just ignore it.
+            if (states.containsKey(state)) {
+                return true;
+            }
+            initialStates.add(state);
+            states.put(state, state);
+            return true;
+        }
+        // TODO build fewer of these
+        Transition transition = new Transition();
+        int totalLeavingState = source.initTransition(currentState, transition);
+        for (int currentLeavingState = 0; currentLeavingState < totalLeavingState; currentLeavingState++) {
+            source.getNextTransition(transition);
+            int min, max;
+            if (transition.max - transition.min >= maxExpand) {
+                // Consider this transition useless.
+                min = 0;
+                max = 0;
+            } else {
+                min = transition.min;
+                max = transition.max;
+            }
+            for (int c = min; c <= max; c++) {
+                codePoints[offset] = c;
+                if (!buildInitial(codePoints, offset + 1, transition.dest)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private void traceRemainingStates() {
+        LinkedList<NGramState> leftToProcess = new LinkedList<NGramState>();
+        leftToProcess.addAll(initialStates);
+        int[] codePoint = new int[1];
+        int statesTraced = 0;
+        Transition transition = new Transition();
+        int currentTransitions = 0;
+        while (!leftToProcess.isEmpty()) {
```

Comment:
Probably should be `false == leftToProcess.isEmpty` given how we write stuff now.